parser_library/src/context/literal_pool.cpp

```diff
@@ -0,0 +1,180 @@
+    const processing::op_code& opcode_ref() const override { return op_code; }
+    processing::processing_format format_ref() const override { return dc_format; }
+    const semantics::operands_si& operands_ref() const override { return op; }
+    const semantics::remarks_si& remarks_ref() const override { return empty_remarks; }
+    const range& stmt_range_ref() const override { return details->r; }
+    const semantics::label_si& label_ref() const override { return empty_label; }
+    const semantics::instruction_si& instruction_ref() const override { return empty_instr; }
+    std::pair<const diagnostic_op*, const diagnostic_op*> diagnostics() const override { return {}; };
+};
+const semantics::remarks_si literal_pool::literal_postponed_statement::empty_remarks({}, {});
+const semantics::label_si literal_pool::literal_postponed_statement::empty_label(range {});
+const semantics::instruction_si literal_pool::literal_postponed_statement::empty_instr(range {});
+const processing::processing_format literal_pool::literal_postponed_statement::dc_format(
+    processing::processing_kind::ORDINARY, processing::processing_form::ASM, processing::operand_occurence::PRESENT);
+
+void literal_pool::generate_pool(dependency_solver& solver, diagnostic_op_consumer& diags)
+{
+    ordinary_assembly_context& ord_ctx = hlasm_ctx.ord_ctx;
+
+    if (m_pending_literals.empty())
+        return;
+
+    for (auto& [it, size, alignment] : m_pending_literals)
+    {
+        const auto& lit = it->first.lit;
+        if (!lit->access_data_def_type()) // unknown type
+            continue;
+
+        size = (semantics::data_def_operand::get_operand_value(*lit, solver).get_length() + 7) / 8;
+        if (size == 0)
+            continue;
+
+        auto top_alignment = size | 16; // 16B length alignment is the top
+        alignment = (~top_alignment & top_alignment - 1) + 1;
+    }
+
+    std::stable_sort(m_pending_literals.begin(), m_pending_literals.end(), [](const auto& l, const auto& r) {
+        return l.alignment > r.alignment;
+    });
+
+    constexpr auto sectalign = doubleword;
+    ord_ctx.align(sectalign);
+
+    for (const auto& [it, size, alignment] : m_pending_literals)
+    {
+        const auto& lit_key = it->first;
+        const auto& lit = lit_key.lit;
+        const auto& lit_val = it->second;
+
+        // TODO: warn on align > sectalign
+        if (size == 0)
+            break;
+
+        bool cycle_ok = ord_ctx.create_symbol(&lit_val.text,
+            ord_ctx.align(lit_val.align_on_halfword ? halfword : no_align),
+            symbol_attributes(symbol_origin::DAT,
+                ebcdic_encoding::a2e[(unsigned char)lit->get_type_attribute()],
+                lit->get_length_attribute(solver)),
+            {});
+
+        ord_ctx.reserve_storage_area(size, no_align);
+
+        if (!cycle_ok)
+            diags.add_diagnostic(diagnostic_op::error_E033(it->second.r));
```

Comment:
test this case?

Reply:
I am reasonably sure that this should never really happen, but not 100%.